Stats: fix wrong paths in page view analytics

vindl opened the pull request:
Resolves https://github.com/Automattic/wp-calypso/issues/23522

According to internal logs, some `calypso_page_view` paths coming from Stats
were reported without replacing passed parameters with corresponding variables,
or removing wrong entries from the route. This was resolved by adding more
restrictive route matching for optional filter and site parameters.

### Testing instructions

1. Set `localStorage.setItem( 'debug', 'calypso:analytics:tracks' )` in the browser console to enable the debug.
2. To reproduce, navigate to `/stats/randomtext` for example.
3. Apply this branch and verify that `/stats/randomtext` is no longer recorded as valid page view path.
4. Try to break it by testing various routes that being with `/stats/`.

Routes to check:
- `/stats`
- `/stats/activity/:site`
- `/stats/day`
- `/stats/insights`
- `/stats/insights/:site`
- `/stats/day/:site`
- `/stats/:module/:site`
- `/stats/day/:module/:site`
- `/stats/post/:post_id/:site`
- `/stats/page/:post_id/:site`
- `/stats/follows/comment/:site`
- `/stats/follows/comment/:page_num/:site`

gwwar commented:
@vindl we probably shouldn't normalize `:period` or `:module`. As @copons noted in https://github.com/Automattic/wp-calypso/pull/23875#issuecomment-378601338 this will likely lead to us missing page views when clicking around in stats without a full page refresh.

I suppose to summarize we should only normalize if it's unknown in code, like an id, user content, or other identifiers like domains.

vindl commented:
> @vindl we probably shouldn't normalize :period or :module. As @Copons noted in #23875 (comment) this will likely lead to us missing page views when clicking around in stats without a full page refresh.

@gwwar `:period` and `:module` are not normalized, I just used that when listing out the routes to be tested to avoid writing every possible combination.

gwwar commented on client/my-sites/stats/controller.jsx:
```diff
@@ -115,24 +112,16 @@ export default {
+	redirectToDefaultSitePage: function( context ) {
 		const siteFragment = getSiteFragment( context.path );
 
-		if ( siteFragment ) {
```
Do we not need to check for this anymore?

vindl replied:
I don't think so, because I want to always redirect when that controller is called - previously it was more of a _maybe_ redirect, which also caused the issue by calling `next()` when `/stats/(.*)` was matched.

gwwar commented:
@vindl yeah sorry I just noticed that while reading through the code :D

gwwar commented:
Tested

- [ ] /stats
- [ ] /stats/activity/:site
- [ ] /stats/:period
- [ ] /stats/insights
- [ ] /stats/insights/:site
- [ ] /stats/:period/:site
- [ ] /stats/:module/:site
- [ ] /stats/:period/:module/:site
- [ ] /stats/post/:post_id/:site
- [ ] /stats/page/:post_id/:site
- [ ] /stats/follows/comment/:site
- [ ] /stats/follows/comment/:page_num/:site

gwwar commented on client/my-sites/stats/activity-log/index.jsx:
```diff
@@ -363,6 +364,7 @@ class ActivityLog extends Component {
 		const requestedRestoreId = this.props.requestedRestoreId || rewindId;
 		return (
 			<div key={ `end-banner-${ restoreId || downloadId }` }>
+				<PageViewTracker path="/stats/activity/:site" title="Stats > Activity" />
```
Oh this needs to be moved under main right? This component file is pretty huge 😅

vindl replied:
Nice catch! :smile: Yeah, I somehow inserted this here instead in the main render method. :man_facepalming:

vindl replied:
Updated in cf356b76e0b54ab3f34140730ca6f42e9afaeff9.

vindl commented:
> The store tab for Jetpack sites is missing a site fragment:

I think that's a separate issue, I was only dealing with routes that start with `/stats` here. I think we have a lot more of those around, due to the pattern to record the result of `sectionify( context.path )` as analytics path, which results in removing the site fragment without properly replacing it with `:site`.

gwwar commented:
Once we get through more of these, I think we can also see if it's possible to 🔪 `sectionify`

yanirs commented:
> @yanirs any concerns here that /stats will always be reported as /stats/day?

Since it looks like `/stats` is the same as `/stats/days`, I don't think it matters.

vindl commented:
@gwwar I updated all of the following items in cf356b76e0b54ab3f34140730ca6f42e9afaeff9:

> `/stats` - This will always report as /stats/day due to the delay. This is a change in behavior, but not sure what's preferred. It's odd that the view change doesn't reflect in the browser url, but it looks like it did this previously.

`/stats` and `/stats/day` was showing the same view, so I added the redirect from `/stats` -> `/stats/day`, which will record page view for `/stats/day`.

> /stats/activity/:site (Note that this is the activity log, so Jetpack sites only)

Fixed by moving the `PageViewTracker` component to correct place in `render` method.

> `/stats/:module/:site` redirects to `/stats` and records `/stats/day`

First I updated this to redirect to `/stats/day/:module/:site`, which is the same view, to conform with `/stats` case. As for the issue itself, I didn't notice this at first while testing since it only happens with clean state. The issue was present in existing code too, and as far as I can see from history, it was a result of sites list removal: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/stats/controller.jsx#L286 (note that `siteId` defaults to `0` a couple of lines above). I'll see if there is some simple solution for this, but it might be a good candidate for separate issue/PR.

> `/stats/follows/comment/:page_num/:site` - looks like we intentionally record as `/stats/follows/comment/:site_id` but if we click on the pager, additional page views are not recorded. In this case we probably do want to normalize page, but we'll run into the same reporting problem. Maybe leave as is and we can circle back later.

I intentionally preserved the existing behavior (removing `:page_num`) here. Would prefer to keep it as is to prevent this PR from blowing up. :)

Unrelated, but I noticed that `DocumentHead` is misbehaving here too (when switching from period to `Weeks` for example). Logged a separate issue here https://github.com/Automattic/wp-calypso/issues/23924. This is a minefield. :smile:

vindl commented:
> I can't remember if I've ever pushed it out and we discussed about it already, or if it was one of my countless discarded experiments, but at some point I tried updating PageViewTracker to make it aware of the URL change instead of the path.

We didn't push it out and the PR was closed, since we figured we can get by without it by using non-normalized `status` . I agree that having a normalized `:page_num` is the right way to go here.

kwight commented:
> I agree that having a normalized :page_num is the right way to go here.

Hm, I think I'd disagree, and rather keep it as is: ignore if there's a page number present, and count 1 page view regardless of going to another page or not.

I consider the purpose of page view analytics is to record the movement and intent of a user through the UI, as opposed to "how did the URL change". I don't find any value separating out whether someone is looking at the first page or the third page, or whether they needed to go from the second to the third. This would be the same existing behaviour as `/comments` too: we track a `calypso_comment_management_change_page` event when changing pages, but not a new page view.

gwwar commented:
> Since we're frequently stumbling upon this kind of occurrences, and we always end up discussing how to work around it, I guess it just makes more sense to change the behaviour of PageViewTracker, and here doing what makes more sense.

Yeah I do think we want to normalize page, but we run into the page view not being recorded. Let's leave that and `/stats/follows/comment/:page_num/:site` for another PR, since I think the changes here are quite an improvement.
